File: library-core/src/main/java/com/mikepenz/fastadapter/IAdapterExtension.java
package com.mikepenz.fastadapter;

import android.os.Bundle;
import android.view.MotionEvent;
import android.view.View;

import java.util.List;

import javax.annotation.Nullable;

/**
 * Created by mikepenz on 04/06/2017.
 */

public interface IAdapterExtension<Item extends IItem> {
    IAdapterExtension<Item> init(FastAdapter<Item> fastAdapter);

    void withSavedInstanceState(@Nullable Bundle savedInstanceState, String prefix);

    void saveInstanceState(@Nullable Bundle savedInstanceState, String prefix);

    boolean onClick(View v, int pos, FastAdapter<Item> fastAdapter, Item item);

    boolean onLongClick(View v, int pos, FastAdapter<Item> fastAdapter, Item item);

    boolean onTouch(View v, MotionEvent event, int position, FastAdapter<Item> fastAdapter, Item item);

    void notifyAdapterDataSetChanged();

    void notifyAdapterItemRangeInserted(int position, int itemCount);

    void notifyAdapterItemRangeRemoved(int position, int itemCount);

    void notifyAdapterItemMoved(int fromPosition, int toPosition);

    void notifyAdapterItemRangeChanged(int position, int itemCount, @Nullable Object payload);

    void set(List<Item> items, boolean resetFilter);

    void performFiltering(@Nullable CharSequence constraint);
}

File: library-core/src/main/java/com/mikepenz/fastadapter/IAdapterExtension.kt
package com.mikepenz.fastadapter

import android.os.Bundle
import android.view.MotionEvent
import android.view.View

/**
 * Kotlin type alias to simplify usage for an all accepting AdapterExtension
 */
typealias GenericAdapterExtension = IAdapterExtension<out GenericItem>

/**
 * Created by mikepenz on 04/06/2017.
 */
interface IAdapterExtension<Item : GenericItem> {
    fun withSavedInstanceState(savedInstanceState: Bundle?, prefix: String)

    fun saveInstanceState(savedInstanceState: Bundle?, prefix: String)

    fun onClick(v: View, pos: Int, fastAdapter: FastAdapter<Item>, item: Item): Boolean

    fun onLongClick(v: View, pos: Int, fastAdapter: FastAdapter<Item>, item: Item): Boolean

    fun onTouch(v: View, event: MotionEvent, position: Int, fastAdapter: FastAdapter<Item>, item: Item): Boolean

    fun notifyAdapterDataSetChanged()

    fun notifyAdapterItemRangeInserted(position: Int, itemCount: Int)

    fun notifyAdapterItemRangeRemoved(position: Int, itemCount: Int)

    fun notifyAdapterItemMoved(fromPosition: Int, toPosition: Int)

    fun notifyAdapterItemRangeChanged(position: Int, itemCount: Int, payload: Any?)

    operator fun set(items: List<Item>, resetFilter: Boolean)

    fun performFiltering(constraint: CharSequence?)
}
